### PR TITLE
[sailfish-browser] Insert new tab next to the parent tab. Fixes JB#30333

### PR DESCRIPTION
**Pull request by rainemak**
When a tab is closed and it had a parent tab, the parent tab activated
otherwise the index next to the previous active tab is activated.

### CODE REVIEW

File: src/declarativewebcontainer.cpp [hunk 1]
```diff
@@ -411,6 +411,14 @@ bool DeclarativeWebContainer::activatePage(const Tab& tab, bool force, int paren
     return false;
 }
 
+int DeclarativeWebContainer::findParentTabId(int tabId)
```
Comment (rojkov) [inline src/declarativewebcontainer.cpp]: This method can be marked as `const`.

File: src/declarativewebcontainer.h [hunk 2]
```diff
@@ -114,6 +114,7 @@ class DeclarativeWebContainer : public QWindow, public QQmlParserStatus, protect
 
     bool isActiveTab(int tabId);
     bool activatePage(const Tab& tab, bool force = false, int parentId = 0);
+    int findParentTabId(int tabId);
```
Comment (rojkov) [inline src/declarativewebcontainer.h]: I guess this also deprecates 
```
int parentTabId(int tabId) const;
```

File: src/declarativetabmodel.cpp [hunk 3]
```diff
@@ -378,6 +378,22 @@ void DeclarativeTabModel::updateActiveTab(const Tab &activeTab, bool loadActiveT
     }
 }
 
+void DeclarativeTabModel::setWebContainer(DeclarativeWebContainer *webContainer)
```
Comment (rojkov) [inline src/declarativetabmodel.cpp]: Just a note: I think there is no need to register tabModels as QML types nowadays thus this method can be dropped with QML type registration altogether.

Comment (rainemak) [inline src/declarativetabmodel.cpp]: Those must be still registered but not as creatable types. Will add a commit on top. Above method is only used from the unit tests at the moment.

**Comment by rainemak** [comment]
Fixed and amended review feedback.

In addition, added a commit on top that removes means to change tab models from qml.

File: src/declarativetabmodel.cpp [hunk 4]
```diff
@@ -161,13 +165,9 @@ void DeclarativeTabModel::closeActiveTab()
 {
     if (!m_tabs.isEmpty()) {
         int index = activeTabIndex();
+        int newActiveIndex = nextActiveTabIndex(index);
         removeTab(m_activeTab.tabId(), m_activeTab.thumbnailPath(), index);
-
```
Comment (rainemak) [inline src/declarativetabmodel.cpp]: Removing last active tab fails. Will fix.

File: src/declarativetabmodel.cpp (hunk 3 above)
Comment (rojkov) [inline src/declarativetabmodel.cpp]: Should this "uncreatableness" be reflected in unit tests as well?

Comment (rainemak) [inline src/declarativetabmodel.cpp]: In tst_webview yes, others are not having webcontainer. Will update.

File: src/declarativetabmodel.cpp [hunk 5]
```diff
@@ -122,31 +128,33 @@ bool DeclarativeTabModel::activateTab(const QString& url)
...
     }
     return false;
 }
 
-bool DeclarativeTabModel::activateTab(int index, bool loadActiveTab)
+void DeclarativeTabModel::activateTab(int index, bool loadActiveTab)
 {
-    if (index >= 0 && index < m_tabs.count()) {
-        const Tab &newActiveTab = m_tabs.at(index);
+    if (m_tabs.isEmpty()) {
```
Comment (rainemak) [inline src/declarativetabmodel.cpp]: Don't do anything if all tabs have been closed.

File: src/declarativetabmodel.cpp [hunk 6]
```diff
@@ -48,10 +50,14 @@ QHash<int, QByteArray> DeclarativeTabModel::roleNames() const
...
 }
 
-void DeclarativeTabModel::addTab(const QString& url, const QString &title) {
+void DeclarativeTabModel::addTab(const QString& url, const QString &title, int index) {
     if (!LinkValidator::navigable(QUrl(url))) {
         return;
     }
+
+    Q_ASSERT(index >= 0);
+    Q_ASSERT(index <= m_tabs.count());
```
Comment (rainemak) [inline src/declarativetabmodel.cpp]: This could be changed to be one-liner

**Comment by rojkov** [comment]
LGTM

**Pull request merged**